feat(complete): generate completions for visible aliases

pzmarzly opened the pull request:
Fixes #4265 

Closes #5412

This PR picks up work from #5412 to get it to a merge-able state.

I'm splitting the work into 2 standalone commits (all tests pass on both of them), to make it easier to review. I leave it up to you whether to merge them in or squash them into one.

Commit 1. Fixing the iteration over `all_subcommands` in `zsh.rs`. We deduplicate values on `(sc_name, bin_name)` keys, but then only iterate on `bin_name`. This doesn't cause problems now, since all bin names seem to be unique. However, without fixing this, the second commit would have started generating duplicated functions with same names.

Commit 2. Starting to generate autocompletions for aliased subcommands. This is taking the code from #5412 and updating it to latest changes.

# Testing the change

Adapting the test code from #4265's first comment:

```toml
[package]
name = "clap-test"
version = "0.1.0"
edition = "2021"
[dependencies]
clap = { path = "../clap", version = "4.5.4" }
clap_complete = { path = "../clap/clap_complete", version = "4.5.2" }
clap_derive = { path = "../clap/clap_derive", version = "4.5.4" }
```

```rust
use std::io::{stdout, Write};
use clap::CommandFactory;
use clap_derive::Parser;
use clap_derive::Args;

#[derive(Parser, Debug)]
enum App {
    #[clap(visible_alias = "bar")]
    Foo(FooArgs),
}

#[derive(Args, Clone, Debug)]
struct FooArgs {
    #[clap(long)]
    my_flag: bool,
}

fn main() {
    let mut app = App::command().disable_help_flag(true).disable_help_subcommand(true);
    let mut buf = vec![];
    clap_complete::generate(clap_complete::Shell::Zsh, &mut app, "clap-test", &mut buf);
    stdout().write_all(&mut buf).unwrap();
}
```

```bash
$ cargo run > before.zsh
$ source before.zsh
$ clap-test [TAB] <- gives me "foo bar --"
$ clap-test foo [TAB] <- gives me "--my-flag"
$ clap-test bar [TAB] <- no reaction
```

After making this change to clap_complete:

```bash
$ cargo run > after.zsh
$ source after.zsh
$ clap-test [TAB] <- gives me "foo bar --"
$ clap-test foo [TAB] <- gives me "--my-flag"
$ clap-test bar [TAB] <- gives me "--my-flag"
```

epage commented on clap_complete/src/shells/zsh.rs:
```diff
@@ -110,12 +110,15 @@ _{bin_name_underscore}_commands() {{
+    let mut all_subcommand_bins: Vec<_> = utils::all_subcommands(p)
+        .into_iter()
+        .map(|(_sc_name, bin_name)| bin_name)
+        .collect();
 
-    all_subcommands.sort();
-    all_subcommands.dedup();
+    all_subcommand_bins.sort();
+    all_subcommand_bins.dedup();
 
-    for (_, ref bin_name) in &all_subcommands {
+    for bin_name in &all_subcommand_bins {
```
> Commit 1. Fixing the iteration over all_subcommands in zsh.rs. We deduplicate values on (sc_name, bin_name) keys, but then only iterate on bin_name. This doesn't cause problems now, since all bin names seem to be unique. However, without fixing this, the second commit would have started generating duplicated functions with same names.

Please update the commit with this information, rather than exclusively leaving it to the PR

Also, could you go into more detail on the problem.  In what way are we getting duplicates with the other PR?

pzmarzly replied:
Updated commit messages, thanks.

If I `git revert` the first commit, then the second one will generate this code:

```zsh
(( $+functions[_my-app__some_cmd_commands] )) ||
_my-app__some_cmd_commands() {
    local commands; commands=(
'sub_cmd:sub-subcommand' \
'help:Print this message or the help of the given subcommand(s)' \
    )
    _describe -t commands 'my-app some_cmd commands' commands "$@"
}
(( $+functions[_my-app__some_cmd_commands] )) ||
_my-app__some_cmd_commands() {
    local commands; commands=(
'sub_cmd:sub-subcommand' \
'help:Print this message or the help of the given subcommand(s)' \
    )
    _describe -t commands 'my-app some_cmd commands' commands "$@"
}
```

Which will work correctly (re-defining a function is okay), but we can fix this inefficiency.

epage commented on clap_complete/src/generator/utils.rs:
```diff
@@ -47,8 +47,15 @@ pub fn subcommands(p: &Command) -> Vec<(String, String)> {
-
         subcmds.push((sc.get_name().to_string(), sc_bin_name.to_string()));
+
+        for alias in sc.get_visible_aliases() {
+            debug!(
+                "subcommands:iter: alias={}, bin_name={}",
+                alias, sc_bin_name
+            );
+            subcmds.push((alias.to_string(), sc_bin_name.to_string()));
```
This only affects zsh and bash.  We need to cover the other shells we support.

pzmarzly replied:
I only edited zsh and bash generators, because only they supported visible_alias at all (by listing aliased subcommands, just not their subcommand).

I just wrote commits that are adding support for visible_alias to fish, elvish and powershell. I have limited access to computers this week, so I cannot test them, but the snapshot tests are looking good.

pzmarzly commented:
Re: linter error. Commit subject max line length of 50 is brutal, even `feat(complete): Autocomplete visible_alias powershell` doesn't fit (53 chars).

pzmarzly commented:
I fixed the linter issues, and implemented support for all shells.

@epage is there something you'd like me to do here, or something I can do to make it easier for you to review it? Many thanks.

epage commented:
Thanks!
